subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -109,38 +132,43 @@ int fota_download_start(char *host, char *file)
 		return -EALREADY;
 	}
 
-	int err = flash_img_init(&flash_img);
+	err = flash_img_init(&flash_img);
 
 	if (err != 0) {
 		LOG_ERR("flash_img_init error %d", err);
 		return err;
 	}
 
-	err = download_client_connect(&dfu, host, &config);
+	bytes_sent_to_flash = 0;
 
-	if (err != 0) {
-		LOG_ERR("download_client_connect error %d", err);
-		return err;
+	return fota_download_from_offset(host, file, 0);
+}
+
+int fota_download_restart(char *host, char *file)
+{
+	if (host == NULL || file == NULL || callback == NULL || bytes_sent_to_flash == 0) {
+		return -EINVAL;
 	}
 
-	err = download_client_start(&dfu, file, 0);
-	if (err != 0) {
-		LOG_ERR("download_client_start error %d", err);
-		download_client_disconnect(&dfu);
-		return err;
+	/* Verify that a download is not already ongoing */
+	if (dfu.fd != -1) {
```

Comment:
I think that if we request a restart, we should forcefully close the connection if one is still valid (`download_client_disconnect`).

Just by looking at this use-case,  when `FOTA_DOWNLOAD_EVT_ERROR` is reported, the connection might still be valid (it's only closed when `download_client` reports an error, not e.g. in case of flash error). In that case we'd fail to ever restart the connection, becase we dont have an API to close it at `fota_*` API level.

Or alternatively, we could handle errors better and make sure the connection is always closed before reporting the error.

Reply:
Good catch...

Reply:
I decided I like the method of having the retry function close the connection if it was still valid.